### DISCUSSION TITLE
Unexpected behavior with atomWithCache, atom keeps re executing when it should not

### DISCUSSION THREAD

**Discussion by Cpt-Falcon** [pmndrs/jotai#2057] [Q&A]
Hi,

I need help working with the atomWithCache. I tried writing an example, but I couldn't easily replicate the issue i'm experiencing. The problem is that it seems like Jotai is not using the cached value and i'm seeing the body of the atomWithCache executing even though it should not. Here's what my atom family generally looks like for reference:

```
export const responseConfigurations = atomFamily(
    ({ guid, tier = null }: { guid: string; tier?: Tier | null }) =>
        atomWithCache(
            (get) => {
                // Content logic, depends on some other atoms as well.

                console.log(guid + ' should not be retriggering');
                return result;
            },
            {
                size: 99999,
                areEqual: (a, b) => {
                    console.log("is equal: " + deepEqual(a, b));
                    return deepEqual(a, b);
                }
            },
        ),
    (x, y) => x.guid === y.guid && x.tier === y.tier,
);
```
I don't see areEqual callback being triggered, but I do see the console.log "should not be retriggering" for some reason even though it should be cached. I'm operating a slider on UI, but that slider is not making any changes to the data flow graph that would affect this response configurations atomFamily. Any thoughts on what could be going on, or any tips on how to mitigate this? 

Here is a screenshot of what the console looks like. I would expect to see the console being triggered in areEqual, and then return false before I see the atomWithCache body being re executed. It just seems like its not being cached. Thanks for your time.

![image](https://github.com/pmndrs/jotai/assets/10540204/1700bd7b-842c-4ec4-a7b9-e9fba12babec)

**Comment by dai-shi** [comment]
https://github.com/jotaijs/jotai-cache/blob/2f153cb3050a74c9a1188158a178015671097f96/src/atomWithCache.ts#L80

Noticed that atomWithCache depends on onMount, and onMount is only invoked when an atom is used (subscribed). #2053
My guess is that you don't use (=useAtomValue) the atom, so it isn't mounted.

It would be nice to implement atomWithCache without onMount, which is probably hacky but more usable.

**Reply by Cpt-Falcon** [reply]
Thanks for your reply. I've copied this hook into my own codebase for testing and you are indeed correct, baseAtom.onMount is being triggered each time, however I am using useAtomValue as you have suggested:

    const currentResponseConfig = useAtomValue(responseConfigurations({ guid: guid }));

It is odd that baseAtom.onMount event is occuring every time, but I will also look for areas in which I am using the atomFamnily without useAtomValue by accident. 

Your response has been very helpful so thank you again.

**Reply by Cpt-Falcon** [reply]
ALso IDK if I would break other stuff with this change, but this seems to fix all of my problems:

![image](https://github.com/pmndrs/jotai/assets/10540204/9a780a39-06c3-419d-93f6-d13d6caa02a4)

I just removed setting writeGetter to null. After all, if its not in the cache its going to read the atom anyways and re add it, so, at least from my lay person perspective, I do not understand the value that setting writegetter to null brings. It seems like you would always want to check the cache, regardless of whether the atom was remounted or not so I do not envision a situation where setting the writegetter to null brings benefits. I totally admit I could have no idea what I'm talking about here and maybe this is a really dumb suggestion.

**Reply by Cpt-Falcon** [reply]
Copilot had a possible answer that sounds sensible, is this correct: "GitHub Copilot: The `baseAtom.onMount` function is called when the `baseAtom` is mounted for the first time. This function is used to set up any necessary state or side effects for the `baseAtom`. In this case, it sets the `writeGetter` value to an array containing a single function that sets the `writeGetter` value to `null`. This is done to ensure that the `writeGetter` value is reset to `null` when the `baseAtom` is unmounted.

The reason why setting `writeGetter` to `null` may prevent unnecessary updates to the cache is because the `useEffect` hook that updates the cache only runs when the `writeGetter` value changes. By setting `writeGetter` to `null`, the hook will not run and the cache will not be updated unnecessarily."

**Reply by dai-shi** [reply]
> baseAtom.onMount is being triggered each time

Shouldn't it be triggered just once?

Can you create a minimal reproduction with https://csb.jotai.org ?

**Reply by Cpt-Falcon** [reply]
@dai-shi Thanks, I will look at replicating what's happening. It may take some time because i'm working on a big application and it may be difficult to pull out the specific elements that are triggering this base atom on mount to trigger.

**Reply by Cpt-Falcon** [reply]
@dai-shi After investigating further I have determined that what's happening is that atom is being deleted and then added again each time my dashboard component re renders. So I have a menu bar on the left with config options, and when I click one of the configs,  there is an updated dashboard with information. During that time, the selected guid atom is updated and the component, which does useAtomValue(selectedGuid), is re rendered accordingly. Then the component passes in the selected guid into the responseConfig atom family to obtain the required responseConfig atom. 

What seems to happen is that one of the atoms in the response config atom family is deleted and then re added. In the jotai debug dev tools I can see the order of the atoms swapping as they are added or removed depending on what I click, and then I see onMount being triggered (which makes sense because the atom was removed and re added, so therefore it would trigger a remount). I'm not doing anything manually or special that adds or removes the atom. 

Is there anything you're aware of that would cause  the atom to be deleted and then re added? 

As far as I understand this should not trigger a remove and then re add right:

    const currentResponseConfig = useAtomValue(responseConfigurations({ guid: guid }));

Based on this new information I will try to replicate it in https://csb.jotai.org

**Reply by Cpt-Falcon** [reply]
Oh I just realized something that may be causing my problems, I am accidentally violating rule #1 of hooks. I'm using useAtom and useAtomValue hooks inside components that are inside loops and conditionals. That is probably causing me significant problems
